Add TriggerMovedTableSchemaCheck

grobyns opened the pull request:
<!-- Thank you for submitting a Pull Request. If you're new to contributing to BCApps please read our pull request guideline below
* https://github.com/microsoft/BCApps/Contributing.md
-->
#### Summary <!-- Provide a general summary of your changes -->
Adds TriggerMovedTableSchemaCheck

#### Work Item(s) <!-- Add the issue number here after the #. The issue needs to be open and approved. Submitting PRs with no linked issues or unapproved issues is highly discouraged. -->
Fixes [AB#523755](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/523755)

nikolakukrika commented on src/Business Foundation/App/NoSeries/src/Upgrade/NoSeriesInstaller.Codeunit.al:
```diff
@@ -34,4 +35,27 @@ codeunit 329 "No. Series Installer"
+#endif
+    begin
+#pragma warning disable AA0175
+        if NoSeries.FindFirst() then;
```
Why do we need this code? First iteration was throwing an error and this iteration is simply reading a row?
Why don't we provide a proper error message except the error of not finding first?
Should we emit telemetry if we cannot find it?

grobyns replied:
it was requested by server as an extra check to load schema metadata just to make sure... the if is there in case the table is empty, which is fine. There's nothing to be done if the code passes. If it fails then it would have failed to sync schema metadata and code execution would stop anyways...

grobyns replied:
Server runtime requested the change as an extra sanity check. The if .. then is there so code doesn't fail when there is no data (the data itself is not relevant). There is no need for an error or telemetry. If the code runs all is good, if it fails it would fail due to a schema bug on the server which would stop all .al code execution anyway.

nikolakukrika commented on src/Business Foundation/App/NoSeries/src/Upgrade/NoSeriesInstaller.Codeunit.al:
```diff
@@ -34,4 +35,27 @@ codeunit 329 "No. Series Installer"
         UpgradeTag.SetUpgradeTag(NoSeriesUpgradeTags.GetImplementationUpgradeTag());
     end;
+
+    local procedure TriggerMovedTableSchemaCheck()
```
The name should be updated that this code is using OnInstall trigger and explain more about the usage of this method.

grobyns replied:
it's a local in an install codeunit... how is that not clear enough?

grobyns replied:
added to upgrade too

nikolakukrika reviewed (dismissed):
If you are implementing this consider:
Moving the check to both OnInstall and Upgrade, you may miss the check otherwise
Name it SchemaSanityCheck or something, add comments why we have this confusing code
Add a method to validate after upgrade is done to check that the data is moved 
Consider emitting telemetry and setting up the alerts 
Have the methods only in v24.x so we track the upgrade, no need to have this in master going forward

grobyns commented:
> If you are implementing this consider: Moving the check to both OnInstall and Upgrade, you may miss the check otherwise Name it SchemaSanityCheck or something, add comments why we have this confusing code Add a method to validate after upgrade is done to check that the data is moved Consider emitting telemetry and setting up the alerts Have the methods only in v24.x so we track the upgrade, no need to have this in master going forward

The check is now on install and on upgrade. Added Sanity to the method name. Added some explanation in a doc trigger.
If we want to remove the code after it has been backported (HF'd) to 24.0 we can but process dictates we do the change in main, then 24.x then 24.0.

darjoo commented on src/Business Foundation/App/NoSeries/src/Upgrade/NoSeriesInstaller.Codeunit.al:
```diff
@@ -34,4 +35,41 @@ codeunit 329 "No. Series Installer"
+    /// </summary>
+    /// <remarks>
+    /// The if .. then statements ensure the code does not fail when the tables are empty. The presence of data is not important, the FindFirst will trigger a schema check.
+    /// Should this code fail it would indicate a bug in the server code. The code is not expected to fail.
```
I have to second this. I understand the server is asking for a second check, but is this a one off? Or is this for every time we move tables? 
If it's one off fine, shouldn't it already have been tested by them? And if it's every time, then clearly something is missing in platform instead.

jehelles replied:
@AndreasMoth 
@grobyns 
@darjoo 
This is a one-off, for the first time, we don't expect any issues, and of course it's tested already (as part of the existing upgrade test, as well as manual testing).

jehelles replied:
You are free to remove it in a month or two.

haoranpb replied:
In that case, should we wrap it in a CLEAN tag?
